tests/graph/test_basic.py

```diff
@@ -550,115 +544,120 @@ def test_get_var_by_name():
-        f2_inp = graph_inputs([f2])
-        assert z not in f2_inp
-        assert x not in f2_inp
-        assert y2 not in f2_inp
-
-    def test_cloning_replace_not_strict_not_copy_inputs(self):
-        # This has nothing to do with scan, but it refers to the clone
-        # function that scan uses internally and that pfunc uses now and
-        # that users might want to use
-        x = vector("x")
-        y = fvector("y")
-        y2 = dvector("y2")
-        z = shared(0.25)
-
-        f1 = z * (x + y) ** 2 + 5
-        f2 = clone_replace(f1, replace=[(y, y2)], strict=False, share_inputs=False)
-        f2_inp = graph_inputs([f2])
-        assert z not in f2_inp
-        assert x not in f2_inp
-        assert y2 not in f2_inp
-
-    def test_clone(self):
-        def test(x, y, mention_y):
-            if mention_y:
-                d = 0.1 + 0 * y
-            else:
-                d = 0.1
-            out = clone_replace(y, replace={x: x + d})
-            return function([], out)()
+        f2 = clone_replace(
+            f1,
+            replace={y: y2} if as_dict else [(y, y2)],
+            strict=strict,
+            share_inputs=share_inputs,
+        )
 
+        f1_inp = [i for i in graph_inputs([f1]) if not isinstance(i, Constant)]
+        f2_inp = [i for i in graph_inputs([f2]) if not isinstance(i, Constant)]
+
+        assert f1 != f2
+
+        # TODO: How is this testing the strict option exactly?  Both cases have
+        # the same assertions.
+        if share_inputs:
+            assert equal_computations([f1], [f2], [y], [y2])
+            assert z in f2_inp
+            assert x in f2_inp
+            assert y2 in f2_inp
+        else:
+            assert not equal_computations([f1], [f2])
+            assert equal_computations([f1], [f2], f1_inp, f2_inp)
+            assert z not in f2_inp
+            assert x not in f2_inp
+            assert y2 not in f2_inp
+
+    @pytest.mark.parametrize("mention_y", [True, False])
+    def test_shared_basic(self, mention_y):
         x = shared(np.asarray(0.0, dtype=config.floatX))
-        utt.assert_allclose(
-            test(x, at.sum((x + 1) ** 2), mention_y=False), 1.21000003815
-        )
-        utt.assert_allclose(
-            test(x, at.sum((x + 1) ** 2), mention_y=True), 1.21000003815
-        )
+        y = at.sum((x + 1) ** 2)
+
+        if mention_y:
+            d = 0.1 + 0 * y
+        else:
+            d = 0.1
+        out = clone_replace(y, replace={x: x + d})
+        res = function([], out)()
+
+        utt.assert_allclose(res, 1.21000003815)
+
+    @pytest.mark.xfail(reason="Default update cloning is not supported.")
+    def test_defaul_updates(self):
+        x = shared(1.0, name="x")
+        x2 = shared(1.0, name="x2")
+        y = scalar("y")
+        y2 = scalar("y2")
+
+        x_update = x + 2
+        x.default_update = x_update
+        # This is only used by `Scan`?
```

Comment:
Seems like `update` preceded `default_update`, and was used as an easy means of passing the updates of RandomVariables when compiling a function, before default_updates became a thing

https://groups.google.com/g/theano-dev/c/1PvvQY8tUHk/m/rDxjUZ9rfHcJ

Reply:
Internally I think it is only used by Scan yes. But I could have missed some use elsewhere

Reply:
The `.update` that `scan` uses appears to be `SymbolicInput.update`, so all traces of the old `Variable.update` might already be gone.

Reply:
I'm starting to think that, if we're going to use an attribute, we should create a subclass of `Variable` that defines the attribute.

Reply:
That sounds reasonable. Would the attribute be mandatory then?

Reply:
Ideally, but we'll need to go over the ways in which `Variable`s are copied/cloned to see if that's sound.

It's also worth considering some other in-graph approaches to update-tracking.